currency_updater.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+"""Currency updater main module."""
+import argparse
+import calendar
+from collections import OrderedDict
+import csv
+from datetime import datetime, timedelta
+from decimal import Decimal
+from pathlib import Path
+import time
+from typing import Any, Callable, TextIO, Tuple, Type, TypeVar, cast
+import urllib.request
+from xml.etree import ElementTree
+
+_BaseDecoratedFunc = Callable[..., Any]  # type: ignore
+DecoratedFunc = TypeVar("DecoratedFunc", bound=_BaseDecoratedFunc)
+
+
+def retry(
+    times: int, exceptions: Tuple[Type[Exception]]
+) -> Callable[[DecoratedFunc], DecoratedFunc]:
+    """Retry Decorator."""
+
+    def decorator(func: DecoratedFunc) -> DecoratedFunc:
+        def newfn(*args: Any, **kwargs: Any) -> Any:  # type: ignore
+            attempt = 0
+            while attempt < times:
+                try:
+                    return func(*args, **kwargs)
+                except exceptions:
+                    print(
+                        f"Exception thrown when attempting to run {func}, attempt "
+                        f"{attempt} of {times}"
+                    )
+                    attempt += 1
+                time.sleep(0.1)
+            return func(*args, **kwargs)
+
+        return cast(DecoratedFunc, newfn)
+
+    return decorator
+
+
+@retry(3, (Exception,))
+def hrmc_urlopen(date: datetime) -> TextIO:
+    """Create a request for the currency exchange rates of the input date."""
+    date_str = date.strftime("%m%y")
+
+    http_headers = {
+        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_8) "
+        "AppleWebKit/537.36 (KHTML, like Gecko) "
+        "Chrome/49.0.2623.112 Safari/537.3"
+    }
+
+    try:
+        return cast(
+            TextIO,
+            urllib.request.urlopen(
+                urllib.request.Request(
+                    "http://www.hmrc.gov.uk/softwaredevelopers/rates/"
+                    f"exrates-monthly-{date_str}.xml",
+                    headers=http_headers,
+                )
+            ),
+        )
+    except urllib.error.HTTPError:
+        return cast(
+            TextIO,
+            urllib.request.urlopen(
+                urllib.request.Request(
+                    "http://www.hmrc.gov.uk/softwaredevelopers/rates/"
+                    f"exrates-monthly-{date_str}.XML",
+                    headers=http_headers,
+                )
+            ),
+        )
```

Comment:
It just retries ignoring exception? I think it's better to log the error and tell that request has failed.